Sphinx documentation expansion for continuous documentation on Read the Docs

chrissimpkins opened the pull request:
This PR expands on the efforts by @gferreira @anthrotype @belluzj @LettError @mashabow to create Sphinx formatted `*.rst` documentation for the fontTools project.  With this PR, all fontTools modules are now included in the Sphinx documentation available in the `Doc` directory and we *should*™ be ready to push to RTD for continuous documentation of the fontTools project.

The PR adds the following:

- new `*.rst` files for modules that were not previously documented with support for automation of source docstring parsing --> formatted documentation with paths organized to match the module paths in the source
- translation of the fontTools README to `*.rst` format for use as the home page (index.html) of the Sphinx documentation (with minor formatting updates)
- Sphinx extension support for local builds of documentation that match the RTD theme (see `Doc/README.md` for install/use details
- Sphinx [Napoleon extension](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/) support for automated parsing of Google Style Guide formatted Python docstrings to documentation files.  This is a much more human legible format than the default Sphinx docstring approach.
- new documentation README in the root of the `Doc` directory with docstring style guidelines, local build instructions, and documentation contributor guidelines
- new CC by-SA 4.0 International License for documentation (defined in `Doc/source/conf.py` file)

I took some liberties with formatting on the `*.rst` doc side but did not modify existing documentation in Python source files or make significant changes to the documentation text used for the translations of existing README text files to this format.  There were no changes to existing documentation in the project README or any other existing documentation files.  These changes support the start of the source documentation work and a great deal of new and modified docstring text will be necessary if we want to use this approach for comprehensive, continuous source documentation.  

For those who would like to build a local set of docs to view these changes, use the following instructions:

(1) setup a development environment for this branch of the fontTools repository and install doc build dependencies with:

```
pip install -e ".[all]"
pip install sphinx sphinx_rtd_theme
```

(2) navigate to the `Doc` directory of the fontTools repository and compile the documentation with one of the two following approaches (based upon availability of `make` on your system):

```
make html
```

or

```
sphinx-build -b html source build
```

The documentation is built on the path `Doc/build/html`.  Open the `index.html` file on that path in your browser to view the home page.

Here are a couple of previews of a local build so that you can see where we are:

### Docs Home

<img width="1327" alt="z40r5-image" src="https://user-images.githubusercontent.com/4249591/47262754-9eece480-d4be-11e8-87f1-c8f4764fd406.png">


### ttLib.ttFont module
![1awdd-image-crunch](https://user-images.githubusercontent.com/4249591/46819675-590f7e00-cd52-11e8-98bb-a6b059f19e06.png)


### TODO:

- [x] add readthedocs yml settings file to specify Python for build, additional optional dependencies necessary to support autoparsing docstrings across all modules in the project https://docs.readthedocs.io/en/latest/yaml-config.html#python-extra-requirements https://github.com/fonttools/fonttools/pull/1333/commits/4b1600223166bc349b09b46439fd8fab0fab8814
- [x] rebase source on master branch when #1335 is merged to master
- [x] add ufoLib documentation when #1335 is merged to master
- [x] add documentation build docs and approach for doc build dependency installs (added new documentation requirements.txt file and build docs)
- [x] update extra dependencies with [ufo] or [all] when #1335 is merged to master
- [x] Is [Google Style Guide docstring formatting](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#381-docstrings) the best approach going forward? (if not, modify the `Doc/README.md`, possibly remove the Napoleon extension from install instructions & `conf.py` settings file)
- [x] Is [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/) an acceptable license for documentation? (if not, modify the `Doc/source/conf.py` field that defines this)
- [x] support Markdown documentation files in addition to rST files? See https://github.com/fonttools/fonttools/pull/1333#discussion_r226404002 (did not add as part of this PR)
- [x] add RTD URL to the `Doc/README.md` file and root `README` file when available

### Active Issues That Have Not Been Addressed

- there are a lengthy series of docstring warnings that are displayed when you `make html` build the documentation.  These will need to be reviewed and addressed as part of future work. See https://github.com/fonttools/fonttools/pull/1333#issuecomment-429575031

Associated issue reports:

- https://github.com/fonttools/fonttools/issues/1266
- https://github.com/fonttools/fonttools/issues/997
- https://github.com/fonttools/fonttools/issues/984

Closes #984

anthrotype commented:
wow Chris, thank you!

anthrotype commented:
Just so you know, I'm in the middle of other things so it'll take some time before i get to review this. I really appreciate the effort though! :clap:

chrissimpkins commented:
No worries Cosimo.  Let me know if acceptable and if you need any changes when you have a chance to review it.

chrissimpkins commented on Doc/README.md:
````diff
@@ -0,0 +1,134 @@
+```
+$ git clone https://github.com/fonttools/fonttools.git
+```
+
````
I need to add venv instructions here.  OK to use the same venv setup instructions that are used on the README or add a link to those instructions?  Any plan to change the dev env setup approach?

anthrotype replied:
don't overcomplicate yourself. Building documentation from source is an advanced user topic. People interested in doing that already know what to do. Just keep it short and simple.

anthrotype replied:
like, you don't need to tell people how to git clone fonttools

chrissimpkins replied:
will simplify.  thank you!

chrissimpkins replied:
fixed in https://github.com/fonttools/fonttools/pull/1333/commits/07f81d968f4e4086817a95921471741671f58f30

chrissimpkins commented:
Note to self:

It looks like we can support installation of the optional Python dependencies required to parse all source files, including those that have optional Python dependency requirements, through a new RTD configuration file:

https://docs.readthedocs.io/en/latest/yaml-config.html#python-extra-requirements

Will look into this.  This would eliminate the the following issues in the original post:

> misc.symfont module documentation does not auto populate with module members using autodoc because the sympy dependency is not installed by default
pens.reportLabPen module documentation does not auto populate with module members using autodoc because reportLab dependency is not installed by default

anthrotype commented:
yes, i was also planning to use extras_require keyword in setup.py. The new fontTools.ufoLib will require fs module, so we'll need to define an 'ufo' extra, so that users can do `pip install fonttools[ufo]` and get the optional dependencies. 
We can do the same for the docs. Let me know what are the exact list of packages needed and I'll add them as "docs" extra to setup.py, the next time I touch it.

chrissimpkins commented on Doc/source/conf.py:
```diff
@@ -25,40 +25,40 @@
 #
 # source_suffix = ['.rst', '.md']
-source_suffix = '.rst'
+source_suffix = ".rst"
```
Associated issue report https://github.com/fonttools/fonttools/issues/1266

I believe that we can convert this to:

```python
source_suffix = ['.rst', '.md']
```

and provide Markdown support in additon to the rST support.  This would allow those who would like to write docs in Markdown to do so.  Unclear to me whether you can include the source parsing scripts in Markdown files.  This may only apply to rST docs?  There are probably situations that warrant free text documentation that is written from scratch. For instance, for the executables that are released with the library.  Those might be good situations for Markdown files.

chrissimpkins replied:
I can confirm that Github flavored Markdown files build to HTML without issues in a `*.rst` + `*.md` Sphinx doc build environment.  Must add the `recommonmark` doc build dependency and the following to the `conf.py` file:

https://github.com/chrissimpkins/fontbakery/blob/11365943499f2f3aa5ee3b3bcb3c0f757095ecf9/docs/source/conf.py#L47-L51

chrissimpkins commented:
> Let me know what are the exact list of packages needed and I'll add them as "docs" extra to setup.py, the next time I touch it.

The ones that I am aware of are:

- sympy
- reportlab

though there are a number of optional Python dependencies listed on the README page and they are all likely going to need to be included in order to pull docstrings from each of those modules.  I wasn't developing this in a venv and suspect that they were being picked up from installed versions on my machine.  Will test it in a venv over the weekend to see if any other errors are raised so that we can address this.

anthrotype commented:
but those two are not "docs" related, strictly speaking.
that reportLabPen.. not even sure who uses that. It's quite niche. I wouldn't worry too much at this stage.
Also, the symfont module is super experimental, and doesn't have a stable API, so I think it can stay outside the docs for now.

chrissimpkins commented:
autodoc must need to import all packages that are used in files in order to parse them.  It raises an error about those missing and does not autopopulate the source members from the modules that include them.  I can remove the documentation files for any experimental / legacy modules if this is what you want.  On the other hand they are part of the current library and it would be helpful to include documentation for the files stating that they are experimental and/or unmaintained legacy code rather than simply having no documentation for them.  Followed by maybe deprecating the legacy stuff and cleaning house in the future?

chrissimpkins commented:
Let me know what direction is best re this issue.  I can add legacy code/deprecation statements in any of the doc files so that we avoid the need to add dependencies to parse docstrings in those modules if this is the preferred approach.

anthrotype commented:
i would prefer you simply skip those for now, yes

chrissimpkins commented:
No problem will do.

chrissimpkins commented:
Full list of autodoc failed import dependency issues when executed from within a new Py3 venv using dev environement install instructions on README (replacing `venv` for `virtualenv` module) and install of doc build dependencies in the new Doc/README.md file here:

```
WARNING: autodoc: failed to import module 'symfont' from module 'fontTools.misc'; the following exception was raised:
No module named 'sympy'

WARNING: autodoc: failed to import module 'reportLabPen' from module 'fontTools.pens'; the following exception was raised:
No module named 'reportlab'

WARNING: autodoc: failed to import module 'plot' from module 'fontTools.varLib'; the following exception was raised:
No module named 'mpl_toolkits'
```

Modules involved:

- fontTools.misc.symfont
- fontTools.pens.reportLabPen
- fontTools.varLib.plot

chrissimpkins commented:
List of other docstring warnings that are raised during the documentation build process (Py3.7 venv):

```
/Users/chris/code/fonttools/Lib/fontTools/cffLib/__init__.py:docstring of fontTools.cffLib.DictCompiler.arg_delta_blend:4: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/cffLib/__init__.py:docstring of fontTools.cffLib.DictCompiler.arg_delta_blend:8: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/cffLib/__init__.py:docstring of fontTools.cffLib.DictCompiler.arg_delta_blend:14: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/cffLib/__init__.py:docstring of fontTools.cffLib.DictCompiler.arg_delta_blend:20: WARNING: Block quote ends without a blank line; unexpected unindent.
<string>:docstring of fontTools.feaLib.ast.BytesIO.seek:7: WARNING: Definition list ends without a blank line; unexpected unindent.
/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/types.py:docstring of fontTools.feaLib.ast.SimpleNamespace:3: WARNING: Inline strong start-string without end-string.
<string>:docstring of fontTools.feaLib.ast.StringIO.seek:7: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/feaLib/builder.py:docstring of fontTools.feaLib.builder.Builder.add_to_cv_num_named_params:1: WARNING: Unknown target name: "self.cv_num_named_params".
/Users/chris/code/fonttools/Lib/fontTools/misc/filenames.py:docstring of fontTools.misc.filenames:9: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/misc/loggingTools.py:docstring of fontTools.misc.loggingTools.configLogger:25: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/misc/loggingTools.py:docstring of fontTools.misc.loggingTools.configLogger:28: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/misc/loggingTools.py:docstring of fontTools.misc.loggingTools.configLogger:30: WARNING: Block quote ends without a blank line; unexpected unindent.
<string>:docstring of fontTools.misc.py23.BytesIO.seek:7: WARNING: Definition list ends without a blank line; unexpected unindent.
<string>:docstring of fontTools.misc.py23.StringIO.seek:7: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/misc/py23.py:docstring of fontTools.misc.py23.open:16: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/misc/py23.py:docstring of fontTools.misc.py23.open:18: WARNING: Unexpected indentation.
/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/types.py:docstring of fontTools.misc.py23.SimpleNamespace:3: WARNING: Inline strong start-string without end-string.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/_c_m_a_p.py:docstring of fontTools.ttLib.tables._c_m_a_p.table__c_m_a_p.getBestCmap:6: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/_c_m_a_p.py:docstring of fontTools.ttLib.tables._c_m_a_p.table__c_m_a_p.getBestCmap:7: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otBase.py:docstring of fontTools.ttLib.tables.otBase.BaseTTXConverter.compile:3: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otBase.py:docstring of fontTools.ttLib.tables.otBase.BaseTTXConverter.compile:6: WARNING: Bullet list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otBase.py:docstring of fontTools.ttLib.tables.otBase.BaseTTXConverter.compile:12: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otTables.py:docstring of fontTools.ttLib.tables.otTables.fixLookupOverFlows:5: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otTables.py:docstring of fontTools.ttLib.tables.otTables.fixLookupOverFlows:8: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otTables.py:docstring of fontTools.ttLib.tables.otTables.fixLookupOverFlows:13: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otTables.py:docstring of fontTools.ttLib.tables.otTables.fixLookupOverFlows:15: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/otTables.py:docstring of fontTools.ttLib.tables.otTables.fixLookupOverFlows:19: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:11: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:13: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:14: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:17: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:24: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:27: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:30: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:33: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:34: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/tables/S_V_G_.py:docstring of fontTools.ttLib.tables.S_V_G_:35: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/ttFont.py:docstring of fontTools.ttLib.ttFont.TTFont.getBestCmap:6: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/ttFont.py:docstring of fontTools.ttLib.ttFont.TTFont.getBestCmap:7: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttLib/ttFont.py:docstring of fontTools.ttLib.ttFont.tagToIdentifier:9: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:17: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:18: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:20: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:22: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:23: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:25: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:29: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:30: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:52: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:60: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:73: WARNING: Unexpected indentation.
/Users/chris/code/fonttools/Lib/fontTools/ttx.py:docstring of fontTools.ttx:75: WARNING: Block quote ends without a blank line; unexpected unindent.

```

chrissimpkins commented:
> yes, i was also planning to use extras_require keyword in setup.py. The new fontTools.ufoLib will require fs module, so we'll need to define an 'ufo' extra, so that users can do pip install fonttools[ufo] and get the optional dependencies.
> We can do the same for the docs. Let me know what are the exact list of packages needed and I'll add them as "docs" extra to setup.py, the next time I touch it.

added in https://github.com/fonttools/fonttools/pull/1333/commits/55d6246e0c04cee41169e8df968ac398ec14f02e.  `pip install -e . [doc]` is now supported for local venv installs + local doc builds.  New Doc/README.md build instructions updated to reflect this change.

chrissimpkins commented:
rebased on master following the #1335 merge of the ufoLib source.  I lost the extras_require [docs] definition in the `setup.py` file during the resolution of merge conflicts.  OK to replace in the new `extras_require` dictionary?  This will cause the doc build dependencies to be installed for anyone who uses `[all]` which may not be what you want.

anthrotype commented:
@chrissimpkins could you remind me what's the status of this and what else needs to be done?

chrissimpkins commented:
The checklist in the OP has a number of the outstanding issues that needed to be addressed at the time of the last edits here.  I will need to rebase again to see where we are with the source changes that have happened since I last looked at this.  I think that the ufoLib merge happened after this work and will need to be included in the documentation files that are part of this PR.  If you are ready to work on it I can check.

davelab6 commented:
Next time Cosimo comes to NYC, Cosimo and Chris should schedule a fonttools
docs hackathon :)

chrissimpkins commented:
Sounds good to me! :)

madig commented:
Side-note: I had weird trouble rebasing your branch onto master, so I did

```
git reset --hard origin/master
git cherry-pick 2eb8b5d3e73e95dfc6e41e43d4b7eb6fb9de3a86^..394fdb5baabe5a1bb3cfb23afeb082df5ba9ed18
```

Only one conflict.

chrissimpkins commented:
Thank you Niko!  I will rebase and push it to my downstream over the weekend.

chrissimpkins commented:
Rebased on master at 9efbb0e74e282c84cfebef8b231027fac9319f03

chrissimpkins commented:
I will have a look at this during the upcoming week to see where we are with the docs vs. the current library modules and determine if there are any other outstanding issues that are not included in the OP checklist.

chrissimpkins commented:
Time available to work on this for the v4.0.0 release?

anthrotype commented:
actually I already pushed the tag for https://github.com/fonttools/fonttools/releases/tag/4.0.0 -- even though Githuh seems to be experiencing a downtime and Travis CI didn't receive the automated request to publish the package to PyPI yet (I hope things will fix themeselves later).
I'll be on holiday next week, and then there's ATypI, so I don't expect to be able to work on this before mid September, sorry. But feel free to rebase and clean up the PR, maybe others will be able to help with the review.

chrissimpkins commented:
Enjoy your vacation!  

No worries.  I will rebase and push.  If anyone has feedback on the questions in the OP please let me know.

chrissimpkins commented:
Rebased on 1b4bb8bc9c8ef4806e96e40f5bf9d573422898b7 and removed `inspect` from the ToC as it appears to have been removed.  If anyone has time to commit to this I will build and run through all of the docs to confirm that there is parity with the current source.  Poke me when the time is right.

madig commented:
What do you mean, what do you need?

chrissimpkins commented:
There is a TODO list in the OP.  A couple of the items require community consensus or at least maintainer blessing:

- [ ] Is Google Style Guide docstring formatting the best approach?
- [ ] Is CC BY-SA 4.0 an acceptable license for documentation?

One is simply a question about the scope of the docs and can be developed down the line:

- [ ] Do we want a documentation guidelines/guidance document similar to Tal's in https://github.com/robofab-developers/fontParts/blob/master/documentation/source/development/documenting.rst?

As of the last attempt to build here there were a few failed import warnings/errors raised on the build attempt which I can look into again.  These may have been resolved since that time.

Lastly, someone with appropriate permissions will need to set up a RTD account and push this as a branch on the repository so that we can see if the docs build there as expected.  Occasionally there is a bit of tinkering necessary to make them ready for release.

BobDu commented:
As a developer, I have been looking forward to fonttools doc for a long time.
So what is the status of this PR now? 
Or has our documentation work made great progress in other PRs?
Can anyone tell me? 
Thanks

chrissimpkins commented:
> So what is the status of this PR now?

I'm available. Any admin available to help me push this across the finish line?

peterennis commented:
Looking for documentation to get started and found this.
It seems to be stuck.
Is there an alternative random script test searches on google?

chrissimpkins commented:
I am going to get together with Cosimo in a few weeks to discuss this PR.  I hope that we will be able to push this to Read The Docs soon!

In the meantime, I believe that you *should* be able to build the documentation files in my branch locally and use them.  See the instructions in the OP.  If you come across any build issues, please post them here and I will address them.

belluzj commented:
Hello all, and thanks a lot for the effort @chrissimpkins! I wonder if I can help make this documentation be available online? Which issues are still blocking this from being merged? I tried just now to generate the documentation on Windows and it works. The contents are not perfect and I saw some warnings during generation but as a starting point it's already great. Tell me what I can do to help.

chrissimpkins commented:
Thanks @belluzj! Do you have admin access to set up the RTD account and start builds there so we can examine the build process for remote build issues? Once we have this information it should be a relatively straightforward update here to include any new source files / refactored paths since this set of changes.

belluzj commented:
No unfortunately I don't think I have those admin rights, I'm not a maintainer. By the way, locally I merged the latest master into your branch and ran the documentation build script again, it went all fine.

chrissimpkins commented:
> No unfortunately I don't think I have those admin rights, I'm not a maintainer.

I wonder if we should fork the repository and push a set of "unofficial" docs on one that we do have admin access on?  I am not excited about maintaining it in that fashion...  This was a significant amount of work and I don't sense that there is admin time available to move it forward.

> By the way, locally I merged the latest master into your branch and ran the documentation build script again, it went all fine.

Good to hear that!

belluzj commented:
To be honest I'm going to put the HTML I just got on some server for internal use at Dalton Maag, but only as a workaround until this PR is merged, because it would be much more practical, as it would keep up-to-date as people contribute to the official repository.

benkiel commented:
>I wonder if we should fork the repository and push a set of "unofficial" docs on one that we do have admin access on?

Please don’t do that, it brings back shudders of the FL Python documentation!

@anthrotype can this be merged/given some priority?

anthrotype commented:
@chrissimpkins thanks Chris, sorry for the long wait. I added you as collaborator to the repo. You should be able to merge the PR and set up the ReadTheDocs now. Let me know if you need anything

benkiel commented:
@chrissimpkins once you are happy with the PR and you have RTD set up, feel free to merge!

twardoch commented:
Well... Since there is the fonttools org, I think that perhaps there could be a repo 

https://github.com/fonttools/fonttools.github.io

which Github would then automatically expose as

https://fonttools.github.io/

Set up the /docs folder there as the place for the hosted docs, and then that place could be used as a place that could pull the automated docs made from the docstrings (in /docs/api for example) but also additional handwritten docs. 

That repo could have different permissions as for who can push, because doc writers are often different people than code writers.

twardoch commented:
@benkiel Fortunately, as we can see from the progress made here, things look much better today than they did 15 years ago — that’s what you get if you share the load and use modern infrastructure!

chrissimpkins commented:
> @chrissimpkins thanks Chris, sorry for the long wait. I added you as collaborator to the repo. You should be able to merge the PR and set up the ReadTheDocs now. Let me know if you need anything

Thanks Cosimo!  I should be able to rebase this and have a detailed look over the weekend.  

> Set up the /docs folder there as the place for the hosted docs, and then that place could be used as a place that could pull the automated docs made from the docstrings (in /docs/api for example) but also additional handwritten docs.

This sounds like an approach that does not use the Read The Docs service for automation of continuous source docstring -> docs builds?

> @chrissimpkins once you are happy with the PR and you have RTD set up, feel free to merge!

Thank you Ben!  Let me get caught up on what is missing from the source documentation since these files were last updated.  I'll be in touch with any questions.

twardoch commented:
Let me amend my suggestion then: 

Set up the fonttools.github.io repo that will give us a simple and nice URL for a web space. The content there should be static plus links. 

It could have a better version of https://github.com/fonttools/fonttools/blob/master/README.rst plus more curated info like simple dumps of the "usage" docs of ttx, pyftsubset etc. and generally content aimed at non-Python developers. 

It could then have simple links to the automated API docs which would then indeed be better at readthedocs, and also links to the wiki etc.

I'll happily co-curate the fonttools.github.io presence. 

This could use Github's own Jekyll service, just like my https://github.com/twardoch/twardoch.github.io which contains just Markdown plus some simple things, and auto-publishes to https://twardoch.github.io/

twardoch commented:
And then both places would serve the docs the best way they're intended: 

- manual content on fonttools.github.io with Github's CI via the simple Jekyll service, aka Github Pages https://pages.github.com/

- auto API content on ReadTheDocs with their CI from docstrings

benkiel commented:
@twardoch These are nice suggestions, but are out of scope for this PR. Please, let's not make the perfect the enemy of the good here, getting this up and live will help many people.

anthrotype commented:
in case you add any new text files that do not get included with the current MANIFEST.in file, make sure to update the latter (that defines the list of files to be included in the source distribution .zip file).
You can build the sdist `python setup.py sdist`, unzip it and build the docs from the unzipped directory to verify you didn't forget any files from the manifest

chrissimpkins commented:
Resolved all open issues and added documentation for modules that were added since I last updated these files.  The RTD account is set up and ready to build.  Will merge to master if CI is OK to confirm that we don't run into build issues on the RTD side.  We should have docs available shortly. 🎉

chrissimpkins commented:
Documentation is now available at https://fonttools.readthedocs.io/en/latest/

peterennis commented:
@chrissimpkins Congratulations on a job well done 👍

belluzj commented:
Thanks a lot! This is awesome

gferreira commented:
congrats & thanks for all the work!

madig commented:
Sphinx has a nice feature where you can cross-link between Sphinx-generated documentation :) I did that for https://ufolib2.readthedocs.io.

chrissimpkins commented:
> Sphinx has a nice feature where you can cross-link between Sphinx-generated documentation :) I did that for https://ufolib2.readthedocs.io.

The internal links are really nice.  There are a handful in spots on the home page.  I noticed earlier today that inspect may no longer exist (or the internal link needs to be updated)?  

The entire set of docs need a once over.  This PR was primarily to build all docstrings that exist, not to update formatting / edit docstrings so that they display properly with sphinx auto doc.  There is definitely work to be done there.

madig commented:
One step at a time :) I think I killed the inspect module some time back because it was a ruin.
